src/main/java/io/jenkins/plugins/mcp/server/Endpoint.java

```diff
@@ -212,306 +221,44 @@ public String getUrlName() {
-     * sends the initial endpoint information to the client.
-     *
-     * @param request  The HTTP servlet request
-     * @param response The HTTP servlet response
-     * @throws IOException If an I/O error occurs
-     */
-    protected void handleSSE(HttpServletRequest request, HttpServletResponse response) throws IOException {
-        if (isClosing.get()) {
-            response.sendError(HttpServletResponse.SC_SERVICE_UNAVAILABLE, "Server is shutting down");
-            return;
+    boolean isStreamableRequest(ServletRequest servletRequest) {
+        if (servletRequest instanceof HttpServletRequest request) {
+            String requestedResource = getRequestedResourcePath(request);
+            return requestedResource.startsWith("/" + MCP_SERVER_STREAMABLE)
+                    && (request.getMethod().equalsIgnoreCase("GET")
+                            || (request.getMethod().equalsIgnoreCase("POST")));
         }
+        return false;
+    }
 
-        response.setContentType("text/event-stream");
-        response.setCharacterEncoding(UTF_8);
-        response.setHeader("Cache-Control", "no-cache");
-        response.setHeader("Connection", "keep-alive");
-        response.setHeader("Access-Control-Allow-Origin", "*");
-
-        String sessionId = UUID.randomUUID().toString();
-        AsyncContext asyncContext = request.startAsync();
-        asyncContext.setTimeout(0);
-
-        PrintWriter writer = response.getWriter();
+    protected void handleSSE(ServletRequest request, ServletResponse response) throws IOException, ServletException {
+        httpServletSseServerTransportProvider.service(request, response);
+    }
 
-        // Create a new session transport
-        HttpServletMcpSessionTransport sessionTransport =
-                new HttpServletMcpSessionTransport(sessionId, asyncContext, writer);
+    protected void handleMessage(HttpServletRequest request, HttpServletResponse response)
+            throws IOException, ServletException {
+        httpServletSseServerTransportProvider.service(new UserContextHttpRequest(objectMapper, request), response);
+    }
 
-        // Create a new session using the session factory
-        McpServerSession session = sessionFactory.create(sessionTransport);
+    private void handleMcpRequest(ServletRequest request, ServletResponse response)
+            throws IOException, ServletException {
         var currentUser = User.current();
         String userId = null;
         if (currentUser != null) {
             userId = currentUser.getId();
         }
-        this.sessions.put(sessionId, new SessionObject(session, userId));
-        var rootUrl = jenkins.model.JenkinsLocationConfiguration.get().getUrl();
-        if (rootUrl == null) {
-            rootUrl = "/";
-        }
-        if (!rootUrl.endsWith("/")) {
-            rootUrl = rootUrl + "/";
+        if (userId != null) {
+            request.setAttribute("userId", userId);
```

Comment:
way better now (previous version with as tool params could be a potential problem) 
But it would be better to use something more specific and not so generic, as it's a servlet request attribute (and why not in a constant somewhere) 
```
    private static final String USER_KEY = Endpoint.class.getName() + ".userId";
```
see javadoc here https://javadoc.io/doc/jakarta.servlet/jakarta.servlet-api/latest/jakarta.servlet/jakarta/servlet/ServletRequest.html